nexus/db-queries/src/db/datastore/silo_group.rs

```diff
@@ -54,39 +281,136 @@ impl DataStore {
         authz_silo: &authz::Silo,
         silo_group: SiloGroup,
     ) -> CreateResult<SiloGroup> {
-        let external_id = silo_group.external_id.clone();
+        let conn = self.pool_connection_authorized(opctx).await?;
+
+        let group_id = silo_group.id();
+        let model: model::SiloGroup = silo_group.into();
+
+        // Check the user's provision type matches the silo
+
+        let silo = {
+            use nexus_db_schema::schema::silo::dsl;
+            dsl::silo
+                .filter(dsl::id.eq(authz_silo.id()))
+                .select(model::Silo::as_select())
+                .get_result_async::<model::Silo>(&*conn)
+                .await
+                .map_err(|e| {
+                    public_error_from_diesel(
+                        e,
+                        ErrorHandler::NotFoundByResource(authz_silo),
+                    )
+                })?
+        };
 
-        DataStore::silo_group_ensure_query(opctx, authz_silo, silo_group)
+        if silo.user_provision_type != model.user_provision_type {
+            error!(
+                self.log,
+                "silo {} provision type {:?} does not match silo group {} \
+                provision type {:?}",
+                authz_silo.id(),
+                silo.user_provision_type,
+                model.id(),
+                model.user_provision_type,
+            );
+
+            return Err(Error::invalid_request(
+                "user provision type of silo does not match silo group",
+            ));
```

Comment:
Might it be helpful to say the two types here like in the above error love?

Reply:
87d5a66c2 changes this to an internal error and adds the two types